### PR TITLE
Plans 2023: Add the featured badges to multiple plans

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to https://github.com/Automattic/martech/issues/1492

## Proposed Changes

Adds the featured badges to multiple plans
- "popular" for premium
- "best for devs" for business
- "best for stores" for eCommerce

~This only applies to the top grid/table, not the comparison grid at the bottom. See the screenshots below for how that looks right now (not great on desktop view, where more than one featured plans sit side-by-side).~

This applies to both the top and bottom (comparison) tables.

**Notes:** The "your plan" badge is handled in another branch/PR (https://github.com/Automattic/wp-calypso/pull/73582).

#### desktop
Screenshot 2023-02-16 at 4 09 13 PM

#### tablet
Screenshot 2023-02-16 at 4 09 33 PM

#### mobile

Screenshot 2023-02-16 at 4 09 50 PM

#### desktop
Screenshot 2023-02-16 at 4 10 16 PM

## Testing Instructions

* go to `/plans/[SITE]` with `?flags=onboarding/2023-pricing-grid` and confirm the changes
* got to `/start/plans` and confirm the same

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plan-features-2023-grid/components/popular-badge.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { getPlanClass } from '@automattic/calypso-products';
+import classNames from 'classnames';
+import PlanPill from 'calypso/components/plans/plan-pill';
+import useFeaturedLabel from '../hooks/use-featured-label';
+
+const PopularBadge: React.FunctionComponent< {
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/components/popular-badge.tsx:6]: This reduces some duplication, keeping things more DRY e.g. the featured labels via the use of a hook. As it stands, it's not complete - some styles should live with their component. Also subject to removal, so remains to be seen.

File: client/my-sites/plan-features-2023-grid/style.scss [hunk 2]
```diff
@@ -698,6 +704,41 @@ body.is-section-signup.is-white-signup,
...
+					&.is-business-plan {
+						border-radius: 0;
+					}
+					&.is-ecommerce-plan {
+						border-top-left-radius: 0;
+					}
+				}
+			}
+		}
+	}
```
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/style.scss:742]: I know there is no better option here but this incredibly painful to see 😬 
We should refactor this table and the sass when the time permits (which means probably never 😆)

**Review by ddc22 — COMMENTED** [review]
I know you are still working on this, but I am leaving my review here. Code looks good but the implementation has some flaws.

image

**Comment by chriskmnds** [comment]
@jdc91 @southp please give this a good round of testing (not code review) if you have cycles, primarily the bottom/ comparison grid. It "should" work as expected. Code-wise, different story.

This is what you should be seeing 🤞 :

Screenshot 2023-02-20 at 12 14 14 AM


These were the parts giving resistance (aligning the borders/outlines top-to-bottom and the rounded corners in the badges):

Screenshot 2023-02-16 at 4 10 16 PM


This pesky little devil almost had me switching careers (perhaps a sign that I should) 😆 :

Screenshot 2023-02-20 at 12 17 09 AM

**Comment by southp** [comment]
Yes, from my testing, that pesky devil in the desktop resolution is the only thing I'm aware of from going through the page in mobile/tablet/desktop in Chrome/Webkit browsers :) 
image

In the plan comparison table it looks well-aligned in both `/start/plans` and `/plans`:
image

File: client/my-sites/plan-features-2023-grid/style.scss (hunk 2 above)
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/style.scss:742]: this table has given me a new perspective on life

**Comment by chriskmnds** [comment]
> Yes, from my testing, that pesky devil in the desktop resolution is the only thing I'm aware of from going through the page in mobile/tablet/desktop in Chrome/Webkit browsers :) image

Thank you, @southp . Do you still see this issue? The top table seems to render fine on my end (Chrome & Firefox). Likely missing something if it pops up on your end 🤔

**Comment by southp** [comment]
@chriskmnds No, I'm no longer seeing that after pulling the latest changes in. There is one last thing though. The border is not aligned in the plans comparison table when there is only one "badged" plan and it's in the last column:
image

Yeah, you are not alone. I too suspect if I should change my career seeing this 😆

File: client/my-sites/plan-features-2023-grid/hooks/use-highlight-adjacency-matrix.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import { useSelector } from 'react-redux';
+import { getCurrentPlan } from 'calypso/state/sites/plans/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import type { PlanProperties } from '../types';
+
+const useHighlightIndices = ( visiblePlans: PlanProperties[] ) => {
+	const selectedSiteId = useSelector( getSelectedSiteId );
+	const currentPlan = useSelector( ( state ) => getCurrentPlan( state, selectedSiteId ) );
+
+	return useMemo( () => {
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/hooks/use-highlight-adjacency-matrix.ts:14]: Pointless using memo here, as any interaction changes `visiblePlans`.

**Comment by chriskmnds** [comment]
Thank you @southp / @jdc91 

> The border is not aligned in the plans comparison table when there is only one "badged" plan and it's in the last column:

yeah, it's been the pattern with this PR. fixing one thing, breaking another..

Feel free to give this another review please, including a code review, if you have cycles. I'll be doing the same tomorrow, so some things may be coming back to clean up. 🙌 

The "your plan" badge is handled in another branch/PR (https://github.com/Automattic/wp-calypso/pull/73582). It took a bit of refactoring to get it in place, so best addressed separately.

**Review by southp — APPROVED** [review]
Thanks for pulling this off. It's tested well and the code LGTM. I do feel that it's a bit too much to acheive what we want, but I think it's already a major step forward to a clearner solution.

Side note: Upon testing, I noticed that the selected plan in the comparison table would jump back to the default value upon resizing:

https://user-images.githubusercontent.com/1842898/220506608-acc18da7-f4a6-413e-9005-577099b26131.mov

However, it's unrelated to this PR so I've documented it as a follow-up after v1: Automattic/martech#1541

**Comment by aneeshd16** [comment]
I confirmed that the logic works well on RTL layouts as well. 👍

**Comment by chriskmnds** [comment]
thanks for testing

> I do feel that it's a bit too much to acheive what we want, but I think it's already a major step forward to a clearner solution.

I am also surprised. it packed quite the challenge to get this right. more than a few iterations if you follow the commit log 🤷‍♂️

**Pull request merged**